### DISCUSSION TITLE
Pointers for Auto Shop Management

### DISCUSSION THREAD

**Discussion by dbob16** [community/community#138076] [Programming Help]
### Body

One project I was thinking about taking up recently is to create an open source shop and inventory management utility. The basic scope of it is the ability to create and print invoices and purchase orders, as well as doing some inventory tasks.

I have some MS Access VBA and Python experience, so not necessarily looking for people to code it for me.

So what I was looking at was using an API like FastAPI to do the network traffic and using a DB that's local to the server like SQLite or Postgres. Since I'm familiar with it I might just use ttk or ttkbootstrap to do the GUI part (I use ttkbootstrap for my tam-py project). Unless ReactJS isn't that bad to pick up.

I have some friends who own mechanic shops and may do a questionnaire of what features they would want to see. Maybe later if the project succeeds I'll see about doing some sort of practical demonstrations and focus groups with local auto shops. I also have connections with CarQuest and our local NAPA chains so I might be able to get it plugged in to be able to price parts on the fly, even if I can't get it to submit PO's/order.

So on top of all of this, is there some other things I should learn on the programming side of things other than API's/FastAPI?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by tpmst** [comment]
Your project sounds very promising, and you already have a solid plan in place! Since you’re already familiar with Python and considering FastAPI for your backend, that’s a great choice for building efficient and scalable APIs. FastAPI is well-suited for asynchronous programming and can handle high volumes of requests, which is beneficial for a shop and inventory management system.

Here are a few additional areas and technologies you might consider learning more about to enhance your project:

Advanced Database Management:

1. Database Design: Since inventory management involves complex relationships, learning about advanced database design and normalization could be beneficial to optimize performance and data integrity.
2. SQLAlchemy: As an ORM (Object-Relational Mapping) tool, SQLAlchemy could be very useful if you choose to work with PostgreSQL. It allows for easier database manipulation through Python code.

Asynchronous Programming:

1. Asyncio: This is Python's library for writing concurrent code using the async/await syntax. It's beneficial when dealing with I/O-bound and high-level structured network code.
2. Starlette: Since FastAPI is built on Starlette, understanding its capabilities could help you better leverage FastAPI’s features for asynchronous support and background tasks.

Frontend Development:

1. ReactJS: If you are considering a web-based frontend, ReactJS is a popular choice due to its efficiency and the vibrant ecosystem around it, including many libraries for UI components that could speed up development. --- You should look into Vue.js or Vite because they have benefits which could help you.
2. Responsive Design: Learning about responsive web design will help ensure that your application looks good on both desktops and mobile devices, which could be useful if users need access from various devices in a shop environment.

Security:

1. OAuth2 with JWT Tokens for Authentication: Since your application will likely be managing sensitive data, understanding how to secure your API using modern authentication techniques like OAuth2 with JWT (JSON Web Tokens) will be important.
